build/Linux+BSD/portable/AppRun.in

```diff
@@ -1,30 +1,50 @@
-#!/bin/bash
-
-# TODO: rewrite this file in a faster, more portable language (e.g. C)
+#!/usr/bin/env bash
 
 # APPDIR and APPIMAGE are environment variables which are set when the user
 # runs the AppImage. They won't be set if you try to run this outside of an
 # AppImage, so here we set them to something sensible instead for testing.
-[  "${APPDIR}"  ] || export APPDIR="$(dirname "$(readlink -f "${0}")")"
-[ "${APPIMAGE}" ] || export APPIMAGE="${APPDIR}/AppRun"
-
-libs="${APPDIR}/lib"
-
-# Workaround for Jack library
-# Based on the solution https://github.com/LMMS/lmms/pull/3958/files
-if ldconfig -p | grep libjack.so.0 > /dev/null 2>&1; then
-   echo "Jack appears to be installed on this system, so we'll use it."
-else
-   echo "Jack does not appear to be installed.  That's OK, we'll use a dummy version instead."
-   export LD_LIBRARY_PATH="${APPDIR}/optional:${LD_LIBRARY_PATH}"
-fi
+[[  "${APPDIR}"  ]] || export APPDIR="$(dirname "$(readlink -f "${0}")")"
+[[ "${APPIMAGE}" ]] || export APPIMAGE="${APPDIR}/AppRun"
+
+function find_library()
+{
+  # Print full path to a library or return exit status 1 if not found
+  local library="$1" # take library name as input (e.g. "libfoo.so")
+  # Look for the library in $LD_LIBRARY_PATH
+  local dir IFS=':' # split path into array on this separator
+  for dir in "${LD_LIBRARY_PATH[@]}"; do
+    if [[ -e "${dir}/${library}" ]]; then
+      echo "${dir}/${library}"
+      return # Library found
+    fi
+  done
+  # Not found yet so check system cache
+  ldconfig -p | awk -v lib="${library}" -v arch="${HOSTTYPE}" \
```

Comment:
- HOSTTYPE [isn't safe](https://www.linuxquestions.org/questions/slackware-14/hosttype-environment-varable-71340/).

- Is `ldconfig -p` output stable?

- Libraries for i386 do not mention architecture in `ldconfig -p` output on my system. Just `(libc6)`. i.e: `libjack.so.0 (libc6) => /usr/lib/i386-linux-gnu/libjack.so.0`

- The running system's architecture is not guaranteed to match the build's target architecture. You should use the build's target architecture here (through CMakeLists.txt I suppose).

Reply:
Maybe you should use `ldd` on the dependencies...

Reply:
`ldd` only shows the libraries that one executable is linked against. It doesn't show:

- Libraries that those libraries link against (dependencies of dependencies)
    - i.e. it is not recursive
- Dependencies that are not explicitly linked against
    - i.e. plugins that are dynamically loaded at runtime via `dlopen`

`ldconfig -p` shows all libraries on the system. Unfortunately this includes libraries for other architectures (if any are installed), hence my attempt to filter the output.

Reply:
> HOSTTYPE isn't safe.

I was aware of potential issues with `i386` vs. `i686` (and I think it gets even worse on ARM), but MuseScore is not producing AppImages for those architectures at the moment, so I thought I'd cross that particular bridge when I came to it.

Same goes for most of the other issues raised, as they also do not apply when the only build target is `x86_64` (nobody runs `x86_64` binaries on non-`x86_64` systems).

> Is `ldconfig -p` output stable?

I've tested on various distros and the output for `x86_64` libraries is always formatted like this:

```
libfoo.so.0 (libc6,x86-64,possible extra info) => /path/to/libfoo.so.0
```

However different distros store their libraries in different locations.

- DEB-based distros seem to use `/lib/x86_64-linux-gnu/`
- RPM-based distros seem to use `/lib64/`

Unfortunately this would break my matching code on RPM systems, because it was matching against the `x86_64` in the path, not the `x86-64` in the parentheses (notice the dash vs. underscore).